server/storage/backend/read_tx.go

```diff
@@ -120,6 +130,106 @@ func (baseReadTx *baseReadTx) UnsafeRange(bucketType Bucket, key, endKey []byte,
 	return append(k2, keys...), append(v2, vals...)
 }
 
+func bytesToRev(bytes []byte) int64 {
+	return int64(binary.BigEndian.Uint64(bytes[0:8]))
+}
+
+func (baseReadTx *baseReadTx) UnsafeRangeWithLock(bucketType Bucket, key, endKey []byte, limit int64, b Backend) ([][]byte, [][]byte) {
+	if endKey == nil {
+		// forbid duplicates for single keys
+		limit = 1
+	}
+	if limit <= 0 {
+		limit = math.MaxInt64
+	}
+	if !bucketType.IsSafeRangeBucket() {
+		panic("do not use UnsafeRangeWithLock on non-keys bucket")
+	}
+
+	if endKey == nil {
+		// single key search, key exists either bolt db or read buffer
+		k1, v1 := baseReadTx.unsafeRangeDB(bucketType, key, endKey, limit)
+		if int64(len(k1)) == limit {
+			return k1, v1
+		}
+		k2, v2 := baseReadTx.unsafeRangeBuff(bucketType, key, endKey, limit-int64(len(k1)), b)
+		return append(k2, k1...), append(v2, v1...)
+	}
+
+	k1, v1 := baseReadTx.unsafeRangeBuff(bucketType, key, endKey, limit, b)
+	if int64(len(k1)) == limit {
+		return k1, v1
+	}
+	k2, v2 := baseReadTx.unsafeRangeDB(bucketType, key, endKey, limit-int64(len(k1)))
+	return append(k1, k2...), append(v1, v2...)
+}
+
+func (baseReadTx *baseReadTx) unsafeRangeBuff(bucketType Bucket, key, endKey []byte, limit int64, b Backend) ([][]byte, [][]byte) {
+	keyRev := bytesToRev(key)
+	endKeyRev := int64(0)
+	if endKey != nil {
+		endKeyRev = bytesToRev(endKey)
+	}
+	var keys [][]byte
+	var vals [][]byte
+	lockHeld := false
+	baseReadTx.mu.RLock()
+	lockHeld = true
+	readBuff := b.ReadTx().UnsafeGetBuffer().(*txReadBuffer)
+	if baseReadTx.buf != readBuff {
+		// backend.readTx.buf has already been updated to a new one by unsafeCommit
+		// batchTxBuffered will write back to the new readTx.buf instead of the baseReadTx.buf
+		// baseReadTx.buf do not need be protected by readTx.mu
+		lockHeld = false
+		baseReadTx.mu.RUnlock()
+	}
```

Comment:
what if we always do RLock here? will it cause any issue? From what I understand, you are basically want to make sure that no one will ever write it back with Lock() then we are giving up RLock(), right?